Load TAs using a user space ELF loader

jenswi-linaro opened the pull request:
This PR introduces a user space ELF loader for TAs.

This PR is rather large so perhaps we should create a separate PR for up to commit "libutee: allocate temp secmem for invoke".

ASLR support still remains.
The script `symbolize.py` needs to be updated to handle the slightly altered dump output. If someone finds it easy to update `symbolize.py` I'll gratefully accept that patch. :-)

jenswi-linaro commented:
Interesting that even though the last two commits has the same commit time, Github is compelled to swap them.

jenswi-linaro commented:
Fixing easy checkpatch errors and pycodestyle errors

jforissier commented:
I tried on HiKey960, i can't load any TA. 32/64 bit doesn't matter.
```
* regression_4002 Test TEE Internal API MAC operations
D/TC:7 0 core_mmu_entry_to_finer_grained:795 xlat tables used 3 / 5
D/TC:? 0 tee_ta_init_pseudo_ta_session:280 Lookup pseudo TA cb3e5ba0-adf1-11e0-998b-0002a5d5c51b
D/TC:? 0 load_ldelf:633 ldelf load address 0x40006000
E/TC:? 0
E/TC:? 0 TA panicked with code 0x0
E/TC:? 0 Status of TA cb3e5ba0-adf1-11e0-998b-0002a5d5c51b (0x3f074a50)
E/TC:? 0  arch: aarch64
E/TC:? 0  region  0: va 0x0000000040000000 pa 0x000000003f000000 size 0x002000 flags ---R-X
E/TC:? 0  region  1: va 0x0000000040002000 pa 0x000000003f077000 size 0x001000 flags ---RW-
E/TC:? 0  region  2: va 0x0000000040004000 pa 0x000000003f200000 size 0x002000 flags rw-RW-
E/TC:? 0  region  3: va 0x0000000040006000 pa 0x000000003f202000 size 0x006000 flags r-x---
E/TC:? 0  region  4: va 0x000000004000c000 pa 0x000000003f208000 size 0x001000 flags rw-RW-
E/TC:? 0 init_with_ldelf:273 ldelf panicked
```

jenswi-linaro commented:
Created PR https://github.com/OP-TEE/optee_os/pull/3052 to take some small patches separated.
I'm considering creating a separate PR with "libutee: allocate temp secmem for invoke" to try to reduce the amount of code to review at a time and also something that is less likely to cause compatibility problems.

There's many corner cases with old and new TAs when it comes to segment ELF layout. I'll try to create a few new test TAs to cover those. The problem reported above is probably due to some old TA with an unexpected load segment layout.

jenswi-linaro commented:
Would you mind posting the output of `readelf -l` for the involved TA and eventual libraries?

jforissier commented:
I can't reproduce the TA load issue. I suppose I must have had a dirty build environment...

I got this assert in ldelf however: https://pastebin.com/BwKgmDUX

jenswi-linaro commented:
Pushed a fix to the assert

jforissier commented:
I'm getting the "ldelf panicked" situation (https://github.com/OP-TEE/optee_os/pull/3036#issuecomment-495522617) again. Taking a closer look.

jforissier commented:
It turns out that "ldelf panicked" is due to an unsupported relocation (`R_AARCH64_ABS64`):
```
$ aarch64-linux-gnu-readelf -r out/arm/ldelf/ldelf.elf

Relocation section '.rela.got' at offset 0x7038 contains 3 entries:
  Offset          Info           Type           Sym. Value    Sym. Name + Addend
000000006020  000000000403 R_AARCH64_RELATIV                    61c8
000000006028  000000000403 R_AARCH64_RELATIV                    6258
000000006030  000000000403 R_AARCH64_RELATIV                    4920

Relocation section '.rela.dyn' at offset 0x7080 contains 3 entries:
  Offset          Info           Type           Sym. Value    Sym. Name + Addend
0000000061e8  000000000403 R_AARCH64_RELATIV                    61d8
0000000061f0  000000000403 R_AARCH64_RELATIV                    61d8
0000000061d0  000300000101 R_AARCH64_ABS64   00000000000061c8 main_elf_queue + 0
```
Can be fixed with:
```diff
diff --git a/ldelf/sub.mk b/ldelf/sub.mk
index 0372da03e5..8c70a4e884 100644
--- a/ldelf/sub.mk
+++ b/ldelf/sub.mk
@@ -1,4 +1,5 @@
 global-incdirs-y += include
+cflags-y := -fvisibility=hidden
 srcs-$(CFG_ARM32_$(sm)) += start_a32.S
 srcs-$(CFG_ARM64_$(sm)) += start_a64.S
 srcs-y += main.c
```

(same as in https://github.com/OP-TEE/optee_os/pull/3035/commits/f00033591dc4934287189db100c8eab1363d500b#r287002307).

jenswi-linaro commented:
Thanks, I've added a fix. How come the relocation only appears sometimes? Depending on compiler version?

jforissier commented:
> How come the relocation only appears sometimes? Depending on compiler version?

Correct. GCC 6.2 generates R_AARCH64_ABS64, GCC 8.3 does not.

jforissier commented:
Now (at ade27a3) the ldelf panic is gone, but `xtest 1019` fails to load the shared library. That is with GCC 6.2 -- all is well with GCC 8.3. I have seen nothing obvious with `readelf`, I guess I should add more debug traces. It is purely related to the shared library (I tried copying the `.ta` file of the shared lib from/to a GCC 6.2/8/3 environment and it breaks or works as expected).

```
root@HiKey960:/ xtest 1019
Test ID: 1019
Run test suite with level=0

TEE test application started with device [(null)]
######################################################
#
# regression
#
######################################################

* regression_1019 Test dynamically linked TA
D/TC:? 0 tee_ta_init_pseudo_ta_session:280 Lookup pseudo TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b
D/TC:? 0 load_ldelf:633 ldelf load address 0x40006000
D/LD:  ldelf:46 Loading TA 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b
D/TC:? 0 tee_ta_init_session_with_context:562 Re-open TA 3a2f8978-5dc0-11e8-9c2d-fa7ae01bbebc
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b (early TA)
D/TC:? 0 system_open_ta_binary:167 res=0xffff0008
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b (Secure Storage TA)
D/TC:? 0 system_open_ta_binary:167 res=0xffff0008
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b (REE [buffered])
D/TC:? 0 system_open_ta_binary:167 res=0x0
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF ffd2bded-ab7d-4988-95ee-e4962fff7154 (early TA)
D/TC:? 0 system_open_ta_binary:167 res=0xffff0008
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF ffd2bded-ab7d-4988-95ee-e4962fff7154 (Secure Storage TA)
D/TC:? 0 system_open_ta_binary:167 res=0xffff0008
D/TC:? 0 system_open_ta_binary:164 Lookup user TA ELF ffd2bded-ab7d-4988-95ee-e4962fff7154 (REE [buffered])
D/TC:? 0 system_open_ta_binary:167 res=0x0
D/TC:? 0 tee_ta_invoke_command:762 Error: ffff0007 of 4
E/LD:  copy_section_headers:701 sys_copy_from_ta_bin
E/TC:? 0 init_with_ldelf:277 ldelf failed with res: 0xffff0007
D/TC:? 0 tee_ta_close_session:488 csess 0x3f074060 id 1
D/TC:? 0 tee_ta_close_session:507 Destroy session
D/TC:? 0 destroy_context:288 Destroy TA ctx (0x3f074b80)
D/TC:? 0 tee_ta_close_session:488 csess 0x3f074400 id 1
D/TC:? 0 tee_ta_close_session:507 Destroy session
E/TC:? 0 tee_ta_open_session:717 Failed. Return error 0xffff0007
/home/jerome/work/optee_build_hikey960/optee_test/host/xtest/regression_1000.c:1421: xtest_teec_open_session(&session, &os_test_ta_uuid, ((void *)0), &ret_orig) has an unexpected value: 0xffff0007 = TEEC_ERROR_BAD_STATE, expected 0x0 = TEEC_SUCCESS
  regression_1019 FAILED
+-----------------------------------------------------
Result of testsuite regression filtered by "1019":
regression_1019 FAILED first error at /home/jerome/work/optee_build_hikey960/optee_test/host/xtest/regression_1000.c:1421
+-----------------------------------------------------
1 subtest of which 1 failed
1 test case of which 1 failed
94 test cases were skipped
TEE test application done!
root@HiKey960:/ 
```

jenswi-linaro commented:
"Update 2" is very interesting. Do you have any clue to what made the section headers to be located that early in the ELF? Compiler/linker version, flags, etc?

jforissier commented:
Apparently, you need GCC 6.2 (not 8.x) to have the sections headers before the 4k boundary, and reverting commit abbf27a27d "TA dev kit: always link shared libraries against OP-TEE libraries" helps as well. The flags are otherwise identical.
Note, this is about the stripped library (libos_test.stripped.so). The headers are always much farther away in the non-stripped file.
The in-kernel ELF loader has no problem reading the section header from a low offset.

jenswi-linaro commented:
Thanks for the information, I'll try to recreate it and fix it.

> The in-kernel ELF loader has no problem reading the section header from a low offset.

That's surprising since that loader (also) assumes that the section headers comes after the last segment. Either way, it doesn't matter, it's just a bit curious.

jenswi-linaro commented:
Pushed a fix for the problem where the section headers are located in the first page of a shared library.

jforissier commented:
@jenswi-linaro 

> Pushed a fix for the problem where the section headers are located in the first page of a shared library.

Yep, works now. `xtest` runs fine in default HiKey960 configuration.

`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960)`

jenswi-linaro commented:
Thanks, I'll rebase this PR once https://github.com/OP-TEE/optee_os/pull/3058 is merged

jforissier commented:
@jenswi-linaro OK, and I still have on my TODO list to adjust `symbolize.py`. Will come soon ;-)

jenswi-linaro commented:
Thanks @jforissier, I was hoping you'd pick that up. :-)

jforissier commented:
> Thanks @jforissier, I was hoping you'd pick that up. :-)

Please see https://github.com/jforissier/optee_os/commits/ldelf-symbolize. I think the two commits would be best folded into "ldelf: support dumping memory map".

jenswi-linaro commented:
Thanks, I've just added that patch as fixup for "ldelf: support dumping memory map"

jforissier commented:
@jenswi-linaro there's one more (https://github.com/jforissier/optee_os/commit/6885f77234de7391056017fe1bd26d572189bced).

jenswi-linaro commented:
Thanks, pushed that one to.

jenswi-linaro commented:
Added a fix where the dynamic segment of a tiny ostest_lib wasn't properly initialized.

jenswi-linaro commented:
Squashed and rebased on top of https://github.com/OP-TEE/optee_os/pull/3058
Added support for ASLR and ftrace

jforissier commented on core/arch/arm/kernel/user_ta.c:
```diff
@@ -1393,3 +1393,34 @@ TEE_Result user_ta_unmap(struct user_ta_ctx *utc, vaddr_t va, size_t len)
+	if (res)
+		return res;
+
+	seg->flags &= ~TEE_MATTR_PROT_MASK;
```
Why? The name `user_ta_set_prot()` sounds like it should only touch the protection bits I would think.

jenswi-linaro replied:
Yes, and that's why we're only clearing those bits here.

jforissier replied:
Duh! You're right of course. Sorry for the noise.

jenswi-linaro commented on ldelf/ta_elf.c:
```diff
@@ -874,6 +923,53 @@ static void print_seg(size_t idx __maybe_unused, int elf_idx __maybe_unused,
 		 idx, width, va, width, pa, sz, flags_str, desc);
 }
 
+static bool get_next_in_order(struct ta_elf_queue *elf_queue,
```
It's almost as you describe it. It's the lowest address which can accomodate the ELF + a random number of pages, the ELF is mapped at the end of the range and the preceding pages are free. This can lead to ELFs being mapped in different orders since a later ELF can fit in a previous range of free pages. I'll add a few words.

jenswi-linaro replied:
```
Adds support in ldelf to load TAs using address space randomization if
CFG_TA_ASLR=y. Works as before with the kernel mode ELF loader with
two exceptions: 
- It falls back to disable ASLR for a particular ELF if it fails to map
  using ASLR.
- Each ELF is loaded with a randomized number of free pages in front.
  These free pages can be reused when loading other ELF permitting
  some ELF to become reordered in memory.

jforissier replied:
Excellent description :) thanks.

jenswi-linaro commented:
Note that the first commits of the PR are the same as in  #3058.
 #3058 was added to try to get some of the early commits of this PR out of the way to make this review easier.

jforissier commented on core/arch/arm/kernel/abort.c:
```diff
@@ -364,7 +364,10 @@ void abort_print_current_ta(void)
 		__print_abort_info(&ai, "User TA");
-	tee_ta_dump_current();
+
+	EMSG_RAW("Status of TA %pUl (%p)",
```
How about moving this text to user space, and drop the context? Arguments "for":
- I don't think the context is useful
- This message logically goes with the subsequent lines (the region dump, the call stack etc.)
- When copying/pasting into `symolize.py`, having a mix of `E/TC` and `E/LD` lines is mildly disturbing (and alignment is not great)
- Anything than can be done in user space should be done there ;-)

For similar reasons, the `arch:` line in `user_ta_dump_state()` would need to move to user space, too.

jenswi-linaro replied:
Agree, but this should be done in "core: dump user ta memory map with ldelf"

jforissier replied:
Totally agree ;-)

jforissier commented on core/arch/arm/pta/system.c:
```diff
@@ -496,6 +496,39 @@ static TEE_Result system_copy_from_ta_binary(struct system_ctx *ctx,
+	if ((flags & accept_flags) != flags)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (flags & PTA_SYSTEM_MAP_FLAG_WRITEABLE)
+		prot |= TEE_MATTR_UW | TEE_MATTR_PW;
```
Why `TEE_MATTR_PW`?

jenswi-linaro replied:
`TEE_MATTR_PW` is redundant as it's implicit by `TEE_MATTR_UW` in the ARM architecture, but I think it's more clear to be explicit.

jforissier commented on lib/libutee/arch/arm/sub.mk:
```diff
@@ -1,11 +1,14 @@
 srcs-$(CFG_ARM32_$(sm)) += utee_mcount_a32.S
 srcs-$(CFG_ARM64_$(sm)) += utee_mcount_a64.S
 
+ifneq ($(sm-$(sm)-is-ld),y)
```
I would prefer `ifneq ($(sm),ldelf)`

jenswi-linaro replied:
I'll fix

jforissier commented on scripts/gen_ldelf_hex.py:
```diff
@@ -0,0 +1,110 @@
+        if segment['p_type'] == 'PT_LOAD':
+            if n == 0:
+                if segment['p_flags'] != (P_FLAGS.PF_R | P_FLAGS.PF_X):
+                    print('Expected first load section to be read/execute')
```
s/section/segment/ (and same below, twice).

jenswi-linaro replied:
I'll fix

jforissier commented on scripts/gen_ldelf_hex.py:
```diff
@@ -0,0 +1,110 @@
+
+    parser.add_argument('--output',
+                        required=True, type=argparse.FileType('wb'),
+                        help='The output dlelf_hex.c')
```
`ldelf_hex.c`

jenswi-linaro replied:
I'll fix

jforissier commented on scripts/gen_ldelf_hex.py:
```diff
@@ -0,0 +1,110 @@
+
+    parser.add_argument('--prefix',
+                        required=True,
+                        help='Prefix used for variables in generated C file')
```
How about hardcoding the prefix and dropping this parameter, since the script is pretty specific to ldelf?

jenswi-linaro replied:
OK, I'll fix.
This is residues from when I had ldelf matching the register width (aarch32/aarch64) of the TAs instead of core.

jenswi-linaro commented on core/arch/arm/pta/system.c:
```diff
@@ -168,11 +186,322 @@ static TEE_Result system_unmap(struct tee_ta_session *s, uint32_t param_types,
+static TEE_Result system_open_ta_binary(struct system_ctx *ctx,
+					uint32_t param_types,
+					TEE_Param params[TEE_NUM_PARAMS])
+{
```
I see your point and agree that's it's a problem, but I think it should be handled at a higher level or I fear that it will become a mess. Which libraries is the TA allowed to read? What about dlopen()?

I think we need a lightweight domain concept where a TA is allowed to read all TAs and libraries within. I'd prefer if I don't have to solve it in this PR, but I can try to add something crude if you think it's important.

jforissier replied:
OK, I think this can wait, because:
1. The GlobalPlatform TEE Protection Profile (v1.2.1) does not seem to mandate confidentiality for TA code (only integrity). In other words, "TA separation" does not seem to encompass this aspect.
2. The vast majority of OP-TEE users must be using REE FS TAs which are not encrypted hence already exposed even to non-secure world.
3. Even if we assume that some users may have developed their own encrypted TA storage, the confidentiality of TA code remains with respect to normal world which is probably the main objective for implementing TA encryption.

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -0,0 +1,811 @@
+	/*
+	 * Add 1Mb pad at end in case a library with this large alignment
+	 * has been mapped before. We want to avoid ending up in a hole in
+	 * the mapping of a library.
```
Why? Many libraries are likely to be less than 1 MB and could fit in that hole.

jenswi-linaro replied:
Because we don't have the patch "core: pta_system: support memory remap" at this stage so we have to stick with the address we get here.

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -0,0 +1,811 @@
+	 * If entry point isn't set explicitly, set to the start of the
+	 * first executable section by the linker. Since ta_head also
+	 * always comes first in legacy TA it means that the entry point
+	 * will be set to 0x20.
```
Perhaps mention when the transition from "legacy" occurred? (although it is not that hard to find).
=> "`Since commit a73b5878c89d ("Replace ta_head.entry with elf entry")`".

jenswi-linaro replied:
I'll fix

jforissier commented on core/kernel/tee_ta_manager.c:
```diff
@@ -868,41 +868,6 @@ struct tee_ta_session *tee_ta_get_calling_session(void)
-
-	dump_state(s->ctx);
-
-	ta_fbuf_dump(s);
```
You forgot this call

jenswi-linaro replied:
Sorry, I must have lost it during one of the many rebases. I'll fix.

jforissier commented on core/arch/arm/kernel/abort.c:
```diff
@@ -31,24 +31,10 @@ enum fault_type {
+	*exidx = 0;
+	*exidx_sz = 0;
+	*stack = 0;
+	*stack_size = 0;
```
The fact that this patch removes the call stack from TA abort dumps, to be re-introduced later, is worth mentioning in the commit description.

jenswi-linaro replied:
I'll fix.

jforissier commented on core/arch/arm/kernel/sub.mk:
```diff
@@ -5,8 +5,6 @@ srcs-$(CFG_EARLY_TA) += early_ta.c
 endif
 srcs-y += pseudo_ta.c
-srcs-y += elf_load.c
-srcs-$(CFG_TA_DYNLINK) += elf_load_dyn.c
```
`CFG_TA_DYNLINK ?= y` should be removed from `mk/config.mk`. Also mention in the commit text that loading dynamically linked TAs is always supported / cannot be disabled.

jenswi-linaro replied:
I'll fix

jenswi-linaro commented on ldelf/include/ldelf.h:
```diff
@@ -0,0 +1,46 @@
+ * In the main C routine the TA is loaded using the PTA System interface.
+ *
+ * When the TA and its dependencies are loaded and relocated ldelf is done
+ * and returns to TEE Core. TEE Core frees the memory allocated to hold ldelf.
```
True, I'll fix

jforissier commented on core/kernel/tee_ta_manager.c:
```diff
@@ -101,6 +101,15 @@ static bool tee_ta_try_set_busy(struct tee_ta_ctx *ctx)
+		 * on for user TAs. Wait here until it's initialized.
+		 */
+		while (ctx->busy)
+			condvar_wait(&ctx->busy_cv, &tee_ta_mutex);
```
Why are we guaranteed to have `ctx->initializing == false` after the wait loop on `ctx->busy`?

jenswi-linaro replied:
Because if `ctx->busy` then `ctx->initializing` will be cleared at the same time as `ctx->busy` is cleared. On the other hand if not `ctx->busy` then we're supposed to continue but with some special treatment in `tee_ta_clear_busy()` since `ctx->flags` has been updated.

jenswi-linaro commented on ldelf/include/ldelf.h:
```diff
@@ -0,0 +1,46 @@
+ * ldelf is loaded into memory by TEE Core. BSS is initialized and a
+ * stack is allocated and supplied in SP register. A struct ldelf_arg
+ * is placed in the stack and a pointer to the struct is provided in
+ * r0/x0.
```
I'm not sure that belongs here. This description is telling what kind of environment `ldelf` can expect. 

It doesn't mention how relevant parts of the ELF is dumped into a C file since that is just how it's implemented in TEE Core.

jforissier replied:
Fair enough :)

jenswi-linaro commented on ldelf/ldelf.mk:
```diff
@@ -0,0 +1,47 @@
+link-out-dir$(sm) := $(out-dir)/$(sm)
+
+cppflags$(sm)	:= $(core-platform-cppflags)
+cflags$(sm)	:= $(core-platform-cflags) -fpie -fvisibility=hidden
```
Some numbers:
```
size ../out-os-qemu/ldelf/ldelf.elf
   text    data     bss     dec     hex filename
  17623     264      16   17903    45ef ../out-os-qemu/ldelf/ldelf.elf

Size of ELF file: 323712
Size of dumped data, sizeof(ldelf_data): 20480

Stripped ELF:
size ../out-os-qemu/ldelf/ldelf.elf
   text    data     bss     dec     hex filename
  17623     264      16   17903    45ef ../out-os-qemu/ldelf/ldelf.elf

Size of ELF file: 26324
Size of dumped data, sizeof(ldelf_data): 20480

Without -ffunction-sections -fdata-sections and -gc
size ../out-os-qemu/ldelf/ldelf.elf
   text    data     bss     dec     hex filename
  20322     264      16   20602    507a ../out-os-qemu/ldelf/ldelf.elf


Size of ELF file: 327464
Size of dumped data, sizeof(ldelf_data): 24576
```

As you can see stripping doesn't affect the size of binary embedded in TEE Core so I guess there's no point in doing it. Dropping -ffunction-sections -fdata-sections and -gc results in a slightly larger embedded binary. I guess it's because our libraries are not optimized for this way of linking (object files has more than one function each etc).

As for `.ARM.extab.text.*`, I'll add an output section.

jforissier replied:
That's fine. Thanks for looking at the numbers that actually matter ;-)

jenswi-linaro replied:
Why do we get `.ARM.extab.text.*` at all? I see that we have those in `tee.elf` also taking up almost 2kB.

jforissier replied:
Sections `.ARM.extab*` contain unwinding information and are complementary `.ARM.exidx*`.

The exception handling spec [1] states that "*An object producer may generate exception-handling table entries (read-only data) in one ELF section, or one section per function. The name of a section containing an exception table fragment must be .ARM.extab optionally followed by further characters.*"

The `-ffunction-sections` flag is what generates one section per function but as the spec says, it's OK to group them in the same output section (which we do for `tee.elf`, and as you suggested, we should do also for `ldelf.elf`).

Not that these tables are not generated when `CFG_UNWIND=n`, which is expected.

[1] http://infocenter.arm.com/help/topic/com.arm.doc.ihi0038b/IHI0038B_ehabi.pdf

jenswi-linaro replied:
It seems that "exception-handling table entries" isn't needed in normal C code as it doesn't support exceptions any way. I assume that exception means the C++ type of exception. We should perhaps in a future optimization discard all the unneeded `.ARM.extab*` sections to save some memory.

jforissier replied:
Agreed.

jenswi-linaro commented on ldelf/link.mk:
```diff
@@ -0,0 +1,53 @@
+
+$(link-out-dir$(sm))/ldelf.elf: $(objs) $(libdeps) $(link-script-pp$(sm))
+	@$(cmd-echo-silent) '  LD      $$@'
+	$(q)$(LD$(sm)) $(ldargs-ldelf.elf) -o $$@
```
See comment at https://github.com/OP-TEE/optee_os/pull/3036#discussion_r297148568

jenswi-linaro commented on ldelf/start_a32.S:
```diff
@@ -0,0 +1,56 @@
+FUNC _ldelf_start , :
+	/*
+	 * First ldelf needs to be relocated. The binary is compiled to
+	 * only contain a .got table which needs to be updated.
```
I'll update the comment.

jenswi-linaro commented on ldelf/start_a32.S:
```diff
@@ -0,0 +1,56 @@
+reloc_begin_rel:
+    .word __reloc_begin - reloc_begin_rel
+reloc_end_rel:
+    .word __reloc_end - reloc_end_rel
```
:-)

jenswi-linaro commented on ldelf/start_a64.S:
```diff
@@ -0,0 +1,59 @@
+FUNC _ldelf_start , :
+	/*
+	 * First ldelf needs to be relocated. The binary is compiled to
+	 * only contain a .got table which needs to be updated.
```
I'll fix.

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -809,3 +814,135 @@ void ta_elf_finalize_mappings(struct ta_elf *elf)
+				     rounddown(seg->vaddr);
+		}
+
+
```
Extra blank line

jforissier commented on ldelf/ta_elf.c:
```diff
@@ -809,3 +814,135 @@ void ta_elf_finalize_mappings(struct ta_elf *elf)
+				/*
+				 * In shared libraries the first page is
+				 * mapped separately with the rest of that
+				 * segment follwing back to back in a
```
"following"

jforissier commented on ldelf/main.c:
```diff
@@ -13,13 +13,21 @@
 #include "sys.h"
 
-static struct ta_elf_queue elf_queue = TAILQ_HEAD_INITIALIZER(elf_queue);
+
```
Do we want this extra line?

jenswi-linaro replied:
I'll fix

jforissier commented on ldelf/sub.mk:
```diff
@@ -5,3 +5,5 @@ srcs-y += main.c
 srcs-y += ta_elf.c
 srcs-y += ta_elf_rel.c
+srcs-y += unwind_arm32.c
+srcs-y += unwind_arm64.c
```
`src-$(CFG_UNWIND)` would be better here (and obviously make `print_stack_arm32()` and `print_stack_arm64()` be `static inline` stubs when `!CFG_UNWIND`).

jenswi-linaro replied:
I'll fix

jenswi-linaro commented on ldelf/ldelf.mk:
```diff
@@ -0,0 +1,47 @@
+# Set current submodule (used for module specific flags compile result etc)
+sm := ldelf
+sm-$(sm) := y
+sm-$(sm)-is-ld := y
```
I'll remove it.

jenswi-linaro commented:
The next push I do to this PR will be a force push with commits fixed up and tags applied, if you don't mind.

jforissier commented:
> The next push I do to this PR will be a force push with commits fixed up and tags applied, if you don't mind.

Sure, please go ahead.

jenswi-linaro commented:
Addressed review comments, tags applied and fixups squashed. Note that a few commit still misses R-B or A-B tages, namely:
- "core: dump ftrace data with ldelf"
- "ldelf: support TA ftrace"
- "ldelf: remap first page of libraries"
- "core: load TAs using ldelf"
- "Add user mode ELF loader"
- "Move elf definition headers to common location"
- "core: remove tee_ta_dump_current()"

Not counting the commits part of #3058

jforissier commented:
For commits:
- "ldelf: remap first page of libraries"
- "core: load TAs using ldelf"
- "Add user mode ELF loader"
- "Move elf definition headers to common location"
- "core: remove tee_ta_dump_current()"

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

I will review the frace commits too, but @b49020 could you please also have a look?

jenswi-linaro commented:
Tags applied, fixed checkpatch warnings/errors that can be fixed (trivial fixes).

b49020 commented:
@jforissier Sure will do.

jforissier commented:
About commit "core: load TAs using ldelf":

Totally removing `CFG_TA_DYNLINK` as I suggested is actually a bit annoying because it was exported to TAs which may depend on it (for example: `xtest 1019` does nothing if the config is not set).
So it would be more correct to have:
```make
$(call force,CFG_TA_DYNLINK,y)
...
ta-mk-file-export-vars-$(sm) += CFG_TA_DYNLINK
```

jenswi-linaro commented:
Reintroduce CFG_TA_DYNLINK

jforissier commented:
For commit "ldelf: support TA ftrace":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

For commit "core: dump ftrace data with ldelf":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960)`

b49020 commented:
@jenswi-linaro I just had a brief look at ftrace dump via `ldelf`. Does it support ftrace dump during TA abort/panic? I could only see dump support during TA close session.

jenswi-linaro commented:
@b49020, isn't the call done in `destroy_session()` enough? AFAICT it's called regardless if the TA panicked or not.

jforissier commented:
@jenswi-linaro @b49020 is right, something must be missing because no ftrace file is generated for e.g. xtest 1019 which ends in a panic :-/

jenswi-linaro commented:
The problem is that the context is reaped early due to the panic/abort and later when the session is closed there's no context left. I'll check what can be done.

b49020 commented:
> isn't the call done in destroy_session() enough?

It seems not to be called during TA aborts like data, prefetch aborts etc. I think you need to extend `abort_print_current_ta` function. Something like:

```
diff --git a/core/arch/arm/kernel/abort.c b/core/arch/arm/kernel/abort.c
index a264462..33db5f4 100644
--- a/core/arch/arm/kernel/abort.c
+++ b/core/arch/arm/kernel/abort.c
@@ -256,6 +256,9 @@ void abort_print_current_ta(void)
                __print_abort_info(&ai, "User TA");
 
        s->ctx->ops->dump_state(s->ctx);
+
+       if (s->ctx->ops->dump_ftrace)
+               s->ctx->ops->dump_ftrace(s->ctx);
 }
 
 static void save_abort_info_in_tsd(struct abort_info *ai)
```

jenswi-linaro commented:
Thanks @b49020, that seems to work. Pushed that fixup.

jforissier commented:
@jenswi-linaro yep, works fine here on HiKey960 :+1:

jenswi-linaro commented on core/kernel/tee_ta_manager.c:
```diff
@@ -278,9 +277,9 @@ static void destroy_session(struct tee_ta_session *s,
 			    struct tee_ta_session_head *open_sessions)
 {
 #if defined(CFG_TA_FTRACE_SUPPORT)
```
This code will be dead code without `CFG_TA_FTRACE_SUPPORT`, but the compiler will not be able to tell that so I'd rather keep the ifdef.

jenswi-linaro commented on core/include/kernel/tee_ta_manager.h:
```diff
@@ -108,9 +107,6 @@ struct tee_ta_session {
 	struct sample_buf *sbuf; /* Profiling data (PC sampling) */
 #endif
-#if defined(CFG_TA_FTRACE_SUPPORT)
-	struct ftrace_buf *fbuf; /* ftrace buffer */
```
Yes, that's easy to add. There is no stable ABI between TEE Core and ldelf. I suggest to add this when needed.

b49020 replied:
Ok, so you suggest to add a new syscall, right?

jenswi-linaro replied:
No, adding another field to `struct ldelf_arg`.

jforissier replied:
> There is no stable ABI between TEE Core and ldelf

And that is quite convenient for us, indeed :smile: :+1:

b49020 replied:
> No, adding another field to struct ldelf_arg.

That's nice.

jenswi-linaro commented:
Rebased on master, squashed fixups, tags applied

b49020 commented:
For commit "ldelf: support TA ftrace" and "core: dump ftrace data with ldelf":

`Reviewed-by: Sumit Garg <sumit.garg@linaro.org>`

jenswi-linaro commented:
@b49020, are you looking at "core: dump ftrace data with ldelf" too or are you done?

b49020 commented:
> are you looking at "core: dump ftrace data with ldelf" too or are you done?

Yes I am done, the tag was for both commits. It seems like I need to rebase my PR: #3080 on top of this work.

jenswi-linaro commented:
Thanks @b49020 
Tags applied

jenswi-linaro commented:
IBART fails due to:
```
1211:  Traceback (most recent call last):
  1212:    File "scripts/gen_ldelf_hex.py", line 12, in <module>
  1213:      from elftools.elf.elffile import ELFFile
  1214:  ImportError: No module named elftools.elf.elffile
```

jenswi-linaro commented:
Hmm, it seems like @jbech-linaro just did something with IBART. :-)

jbech-linaro commented:
> Hmm, it seems like @jbech-linaro just did something with IBART. :-)

I apt-get'd the python packages and restarted the job.

jenswi-linaro commented:
The IBART error is supposed to be fixed with https://github.com/OP-TEE/optee_os/pull/3101

jbech-linaro commented:
> The IBART error is supposed to be fixed with #3101

Try to restart the build **after** the ongoing #3101 IBART job has been merged.

jbech-linaro commented:
> @jbech-linaro
> 
> > Try to restart the build **after** the ongoing #3101 IBART job has been merged.
> 
> Does IBART merge the PR with the current tip of the master branch before building?

Ah, sorry, you're right, Jens would of course need to rebase #3101, since the patch/PRs are touching the same git. IBART picks the branch verbatim from the PR creator.

jforissier commented:
OK thanks for clarifying. I am fine with this behavior actually, because merging automatically would risk causing conflicts at any moment which could be annoying in the review process.

jenswi-linaro commented:
Rebased on master now that #3101 is merged.

jforissier commented:
The checkpatch warning about SPDX in "ldelf: support dumping TA call stack" seems legit.

jenswi-linaro commented:
So what do you want there?

jforissier commented:
Errr... I don't know :-/

Scancode [1] identifies the license as "Carnegie Mellon Contributors" but there is no SPDX identifier for this type of license apparently. So let's ignore the warning.

[1] https://github.com/nexB/scancode-toolkit

jforissier commented:
Now the IBART error is due to lockdep not supporting `mutex_trylock()`. I'll see if I can come up with a fix.

jforissier commented:
https://github.com/OP-TEE/optee_os/pull/3103

jenswi-linaro commented:
Rebased on master now that #3103 is merged.

jforissier commented:
I think this PR is now ready for merging, thanks!
